src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs

```diff
@@ -183,112 +197,164 @@ protected virtual string GenerateUniqueId()
+                        // attributes have string values and may be calculated using a temporary TextWriter or other
+                        // buffer.
+                        var bufferableAttribute = boundStringAttributeNames.Contains(attributeDescriptor.Name);
 
-                    // If we haven't recorded a value and we need to buffer an attribute value and the value is not
-                    // plain text then we need to prepare the value prior to setting it below.
-                    if (!attributeValueRecorded && bufferableAttribute && !isPlainTextValue)
+                        RenderNewAttributeValueAssignment(
+                            attributeDescriptor,
+                            bufferableAttribute,
+                            attributeValueChunk,
+                            valueAccessor);
+
+                        // Execution contexts are a runtime feature.
+                        if (_designTimeMode)
+                        {
+                            continue;
+                        }
+
+                        // We need to inform the context of the attribute value.
+                        _writer
+                            .WriteStartInstanceMethodInvocation(
+                                ExecutionContextVariableName,
+                                _tagHelperContext.ExecutionContextAddTagHelperAttributeMethodName)
+                            .WriteStringLiteral(attributeDescriptor.Name)
+                            .WriteParameterSeparator()
+                            .Write(valueAccessor)
+                            .WriteEndMethodInvocation();
+                    }
+                    else
                     {
-                        BuildBufferedWritingScope(attributeValueChunk, htmlEncodeValues: false);
+                        // The attribute value has already been recorded, lets retrieve it from the stored value
+                        // accessors.
+                        _writer
+                            .WriteStartAssignment(valueAccessor)
+                            .Write(htmlAttributeValues[attributeDescriptor.Name])
+                            .WriteLine(";");
                     }
+                }
+            }
 
-                    // We capture the tag helpers property value accessor so we can retrieve it later (if we need to).
-                    var valueAccessor = string.Format(CultureInfo.InvariantCulture,
-                                                      "{0}.{1}",
-                                                      tagHelperVariableName,
-                                                      attributeDescriptor.PropertyName);
+            // Pass 3: Add all remaining attributes that match a prefix to corresponding dictionary. Use same order as
+            // above loops but skip everything already bound to a specific property of this tag helper.
+            foreach (var attributeDescriptor in attributeDescriptors)
```

Comment:
When doing #225, will still do Pass 1.  Passes 2 and 3 would instead be a per-tag-helper precedence applied while going through the attributes.

Alternatively could add a Pass 4 for #225: Reorder the attribute lists created earlier to match the source order.